File: kotlin/plugins/Secretspec.kt
// SecretSpec, as a voxgig/plugin definition.

package com.voxgig.sekreto.plugins

import com.voxgig.sekreto.Definition
import com.voxgig.sekreto.Provider
import com.voxgig.sekreto.SekretoError
import com.voxgig.sekreto.envkey
import com.voxgig.sekreto.providerplugin

/**
 * SecretSpec (https://secretspec.dev).
 *
 * SecretSpec is a declaration - a `secretspec.toml` naming the secrets a
 * project needs - plus a chain of its own backends to satisfy them from.
 * That makes it the same shape as sekreto one level down, and the reason
 * to support it is the same reason sekreto exists: a project that has
 * already declared its secrets there should not have to declare them
 * again here.
 *
 * Read through its CLI, as boru is, because that is the interface it
 * offers a program in another language: `secretspec get API_TOKEN`
 * prints the value on stdout and nothing else. A sekreto name maps to a
 * SecretSpec key exactly as it maps to an environment variable -
 * `api.token` is `API_TOKEN` - which is the convention SecretSpec's own
 * examples use.
 *
 * `backend` selects one of SecretSpec's backends (`--provider`, e.g.
 * `keyring` or `dotenv://.env`) and is called `backend` here only
 * because `provider` already means something else in this library.
 *
 * A reason is required, not optional: SecretSpec records every read in
 * an audit log and refuses to read at all without one. sekreto sends
 * `sekreto` unless told otherwise, so the audit trail says which tool
 * asked.
 */
class Secretspec(
    command: String? = null,
    private val file: String? = null,
    private val profile: String? = null,
    private val backend: String? = null,
    private val reason: String? = null,
    private val prefix: String? = null,
) : Provider {

    private val command: String =
        if (command.isNullOrEmpty()) "secretspec" else command

    override fun lookup(name: String): String? {
        val key = envkey(name, prefix)

        val args = mutableListOf(command)
        if (!file.isNullOrEmpty()) {
            args.add("--file")
            args.add(file)
        }
        args.add("get")
        args.add(key)
        if (!backend.isNullOrEmpty()) {
            args.add("--provider")
            args.add(backend)
        }
        if (!profile.isNullOrEmpty()) {
            args.add("--profile")
            args.add(profile)
        }
        args.add("--reason")
        args.add(first(reason, "sekreto"))

        val (out, why, status) = runcmd(ProcessBuilder(args), command)

        if (0 == status) {
            // The value and one newline, and nothing else.
            return out.removeSuffix("\n")
        }

        if (secretspecmiss(why, key)) {
            return null
        }

        throw SekretoError(
            "sekreto: secretspec error: " + why.ifEmpty { "exit $status" },
        )
    }

    override fun describe(): String =
        "secretspec" + if (backend.isNullOrEmpty()) "" else ":$backend"
}

/**
 * Does this SecretSpec failure mean "no such secret" rather than "I could
 * not answer"?
 *
 * SecretSpec says `Secret 'API_TOKEN' not found` for both a name it does
 * not declare and one declared with no value, and both are misses: this
 * store does not hold it, so the chain carries on.
 *
 * MATCHED ON THE WHOLE PHRASE, NOT ON "not found". SecretSpec also says
 * `Provider backend 'keyring' not found`, which is a store that could not
 * answer at all - and reading that as a miss is the worst failure this
 * library has, because the chain then falls through to a weaker store
 * without saying so. The key is required to appear, so the two cannot be
 * confused.
 */
internal fun secretspecmiss(why: String, key: String): Boolean =
    why.contains("Secret '$key' not found")

/** The `secretspec` provider kind, as a voxgig/plugin definition. A
 * consumer imports this and hands it to `Sekreto`; a consumer that does
 * not is a `Sekreto` that cannot build a `secretspec` chain entry. */
val secretspec: Definition = providerplugin("secretspec") { spec ->
    Secretspec(
        spec.command, spec.file, spec.profile, spec.backend, spec.reason, spec.prefix,
    )
}

File: scala/plugins/Secretspec.scala
// SecretSpec, as a voxgig/plugin definition.

package com.voxgig.sekreto.plugins

import scala.jdk.CollectionConverters.*

import com.voxgig.sekreto.*

/** SecretSpec (https://secretspec.dev).
  *
  * SecretSpec is a declaration - a `secretspec.toml` naming the secrets a
  * project needs - plus a chain of its own backends to satisfy them from.
  * That makes it the same shape as sekreto one level down, and the reason
  * to support it is the same reason sekreto exists: a project that has
  * already declared its secrets there should not have to declare them
  * again here.
  *
  * Read through its CLI, as boru is, because that is the interface it
  * offers a program in another language: `secretspec get API_TOKEN` prints
  * the value on stdout and nothing else. A sekreto name maps to a
  * SecretSpec key exactly as it maps to an environment variable -
  * `api.token` is `API_TOKEN` - which is the convention SecretSpec's own
  * examples use.
  *
  * `backend` selects one of SecretSpec's backends (`--provider`, e.g.
  * `keyring` or `dotenv://.env`) and is called `backend` here only because
  * `provider` already means something else in this library.
  *
  * A reason is required, not optional: SecretSpec records every read in an
  * audit log and refuses to read at all without one. sekreto sends
  * `sekreto` unless told otherwise, so the audit trail says which tool
  * asked.
  */
class Secretspec(
    commandgiven: Option[String] = None,
    file: Option[String] = None,
    profile: Option[String] = None,
    backend: Option[String] = None,
    reason: Option[String] = None,
    prefix: Option[String] = None,
) extends Provider:

  private val command: String = commandgiven.filter(_.nonEmpty).getOrElse("secretspec")

  override def lookup(name: String): Option[String] =
    val key = envkey(name, prefix)

    val args = scala.collection.mutable.ListBuffer(command)
    file.filter(_.nonEmpty).foreach(value => args ++= List("--file", value))
    args ++= List("get", key)
    backend.filter(_.nonEmpty).foreach(value => args ++= List("--provider", value))
    profile.filter(_.nonEmpty).foreach(value => args ++= List("--profile", value))
    args ++= List("--reason", first(reason, Some("sekreto")))

    val ran = runcmd(ProcessBuilder(args.asJava), command)

    if 0 == ran.status then
      // The value and one newline, and nothing else.
      Some(dropsuffix(ran.out, "\n"))
    else if secretspecmiss(ran.why, key) then None
    else
      throw SekretoError(
        "sekreto: secretspec error: " +
          (if ran.why.isEmpty then s"exit ${ran.status}" else ran.why),
      )

  override def describe(): String =
    "secretspec" + (if backend.exists(_.nonEmpty) then s":${backend.get}" else "")

/** Does this SecretSpec failure mean "no such secret" rather than "I could
  * not answer"?
  *
  * SecretSpec says `Secret 'API_TOKEN' not found` for both a name it does
  * not declare and one declared with no value, and both are misses: this
  * store does not hold it, so the chain carries on.
  *
  * MATCHED ON THE WHOLE PHRASE, NOT ON "not found". SecretSpec also says
  * `Provider backend 'keyring' not found`, which is a store that could not
  * answer at all - and reading that as a miss is the worst failure this
  * library has, because the chain then falls through to a weaker store
  * without saying so. The key is required to appear, so the two cannot be
  * confused.
  */
private[plugins] def secretspecmiss(why: String, key: String): Boolean =
  why.contains(s"Secret '$key' not found")

/** The `secretspec` provider kind, as a voxgig/plugin definition. A consumer
  * imports this and hands it to `Sekreto`; a consumer that does not gets a
  * `Sekreto` that cannot build a chain entry of that kind.
  */
val secretspec: Definition = providerplugin("secretspec", spec =>
  Secretspec(spec.command, spec.file, spec.profile, spec.backend, spec.reason, spec.prefix))
